src/main/java/com/oneindexed/spring/template/SpringTemplateDefinitionParser.java

```diff
@@ -0,0 +1,57 @@
+package com.oneindexed.spring.template;
+
+import org.springframework.beans.factory.support.AbstractBeanDefinition;
+import org.springframework.beans.factory.support.BeanDefinitionBuilder;
+import org.springframework.beans.factory.support.BeanDefinitionValidationException;
+import org.springframework.beans.factory.xml.AbstractBeanDefinitionParser;
+import org.springframework.beans.factory.xml.ParserContext;
+import org.springframework.lang.Nullable;
+import org.springframework.util.StringUtils;
+import org.w3c.dom.Element;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SpringTemplateDefinitionParser extends AbstractBeanDefinitionParser {
+
+    private static final String PROPERTY_REF = "property-ref";
+    private static final String PROPERTY_PROVIDER = "property-provider";
+    private static final List<String[]> ATTRIBUTES = new ArrayList<>();
+
+    static {
+        ATTRIBUTES.add(new String[] {"template", "template"});
+        ATTRIBUTES.add(new String[] {PROPERTY_REF, "propertyRef"});
+        ATTRIBUTES.add(new String[] {PROPERTY_PROVIDER, "propertyProvider"});
+        ATTRIBUTES.add(new String[] {"lazy-init", "lazyInit"});
+        ATTRIBUTES.add(new String[] {"persistent-definitions", "persistentDefinitions"});
+    }
+
+    @Nullable
+    protected AbstractBeanDefinition parseInternal(Element element, ParserContext parserContext) {
+
+        BeanDefinitionBuilder builder = BeanDefinitionBuilder
+                .rootBeanDefinition(SpringTemplateBean.class);
+
+        // Map from the xml style name to the java style name
+        ATTRIBUTES.forEach(attribute -> {
+            String attributeValue = element.getAttribute(attribute[0]);
+            if (StringUtils.hasText(attributeValue)) {
+                builder.addPropertyValue(attribute[1], attributeValue);
+            }
+        });
+
+        // Fast fail
```

Comment:
Fast fail doesn't make sense occurring after we've created the builder and set some of the properties - move it to the beginning of the method

Reply:
Done